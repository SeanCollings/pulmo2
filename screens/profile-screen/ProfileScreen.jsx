import React, { useContext, useEffect, useState } from 'react';
import {
  StyleSheet,
  View,
  Text,
  TouchableOpacity,
  ActivityIndicator,
} from 'react-native';
import { MaterialCommunityIcons } from '@expo/vector-icons';

import options from './options';
import { COLORS } from '../../constants/constants';
import { HistoryContext } from '../../context/history-context';
import {
  convertDate,
  getRemainingTime,
  getTotalResultTime,
  getDay,
} from '../../utils';
import ListScroller from '../../components/ListScroller';
import { STRENGTH_KEY, ENDURANCE_KEY } from '../../data';
import LevelSelector from '../../components/LevelSelector';

export const profileScreenOptions = options;

const getIcon = (type) => {
  if (type === STRENGTH_KEY) {
    return 'dumbbell';
  } else if (type === ENDURANCE_KEY) {
    return 'clock';
  } else {
    return 'wrench';
  }
};

const RenderItemHistory = ({ item, navigation }) => {
  const { date, excercise, results, type, level } = item;
  const totalTime = getTotalResultTime(results);
  const icon = getIcon(type);

  return (
    <View style={styles.activityItemContainer}>
      <TouchableOpacity
        style={styles.activityResultContainer}
        onPress={() => navigation.navigate('Activity', { item })}
        activeOpacity={0.6}
      >
        <View style={styles.iconContainer}>
          <MaterialCommunityIcons name={icon} color={COLORS.BORDER} size={30} />
        </View>
        <View style={styles.activityTextContainer}>
          <Text style={styles.activityHeading}>{excercise.title}</Text>
          <Text style={styles.activityText}>
            {getDay(date)} {convertDate(date)}
          </Text>
          <View style={styles.bottomItemContainer}>
            <Text style={{ ...styles.activityText, opacity: 0.7 }}>
              {getRemainingTime(totalTime)}
            </Text>
            <Text style={{ ...styles.activityText, opacity: 0.7 }}>
              level: {level}
            </Text>
          </View>
        </View>
      </TouchableOpacity>
      <View style={{ ...styles.verticalLine }}></View>
    </View>
  );
};

const ProfileScreen = ({ navigation }) => {
  const { activities, getSavedActivites } = useContext(HistoryContext);
  const [allActivites, setAllActivites] = useState(activities);
  const [isLoading, setIsLoading] = useState(true);

  useEffect(() => {
    if (!activities.length) {
      getSavedActivites().then(() => {
        setIsLoading(false);
      });
    } else {
      setIsLoading(false);
    }
  }, []);

  useEffect(() => {
    const sortedActivites = activities.sort(
      (a, b) => new Date(b.date) - new Date(a.date)
    );

    setAllActivites(sortedActivites);
  }, [activities]);

  return (
    <View style={styles.container}>
      <LevelSelector header="difficulty level" />
      <View style={styles.historyContainer}>
        <Text style={styles.historyHeaderText}>History</Text>
      </View>
      <View style={styles.verticalLine}></View>
      <View style={styles.listContainer}>
        {isLoading && (
          <View style={{ paddingTop: 20 }}>
            <ActivityIndicator size="large" color={COLORS.SECONDARY} />
          </View>
        )}
        {!isLoading && (
          <ListScroller
            data={allActivites}
            uid={'date'}
            RenderItem={RenderItemHistory}
            navigation={navigation}
          />
        )}
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    backgroundColor: COLORS.BACKGROUND,
  },
  verticalLine: {
    width: '80%',
    borderTopColor: COLORS.BORDER,
    borderTopWidth: 1,
    opacity: 0.6,
  },
  historyContainer: { paddingVertical: 10, alignItems: 'center' },
  historyHeaderText: {
    fontSize: 20,
    fontFamily: 'tit-regular',
    color: COLORS.SECONDARY,
    textTransform: 'lowercase',
  },
  listContainer: {
    flex: 1,
    paddingTop: 5,
    paddingBottom: 20,
  },
  activityItemContainer: {
    alignItems: 'center',
    paddingTop: 5,
  },
  activityResultContainer: {
    flexDirection: 'row',
    paddingBottom: 5,
    paddingHorizontal: '10%',
  },
  iconContainer: {
    alignItems: 'center',
    width: '20%',
    justifyContent: 'center',
  },
  activityTextContainer: {
    width: '80%',
  },
  activityHeading: {
    fontFamily: 'tit-regular',
    fontSize: 18,
    color: COLORS.SECONDARY,
  },
  activityText: {
    fontFamily: 'tit-light',
    fontSize: 14,
    color: COLORS.TEXT,
  },
  bottomItemContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    width: '100%',
    paddingRight: 10,
  },
});

export default ProfileScreen;
