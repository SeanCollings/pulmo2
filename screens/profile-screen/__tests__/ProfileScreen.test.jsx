import React from 'react';
import { act, create } from 'react-test-renderer';
import { cleanup, fireEvent } from 'react-native-testing-library';
import ProfileScreen, { profileScreenOptions } from '../ProfileScreen';
import ProfileContextProvider from '../../../context/profile-context';

jest.useFakeTimers();
jest.mock('react-native/Libraries/Animated/src/NativeAnimatedHelper');

describe('ProfileScreen - unit test', () => {
  let component;

  afterEach(() => {
    cleanup();
    jest.clearAllMocks();
  });

  test('should render', async () => {
    await act(async () => {
      component = create(
        <ProfileContextProvider>
          <ProfileScreen />
        </ProfileContextProvider>
      );
    });

    expect(component).toMatchSnapshot();
  });

  describe('profileScreenOptions', () => {
    test('options', () => {
      const options = profileScreenOptions();
      expect(options).toMatchSnapshot();
    });

    test('headerTitle', () => {
      const headerTitle = profileScreenOptions().headerTitle({
        tintColor: 'white',
      });
      expect(headerTitle).toMatchSnapshot();
    });

    test('headerRight', () => {
      const headerRight = profileScreenOptions().headerRight({
        tintColor: 'white',
      });
      expect(headerRight).toMatchSnapshot();
    });

    test('should handle headerRight navigate to Settings', async () => {
      const navigate = jest.fn();
      const navData = { navigation: { navigate } };

      await act(async () => {
        component = create(
          profileScreenOptions(navData).headerRight({
            tintColor: 'white',
          })
        );
      });

      const headerButton = component.root.find(
        ({ props }) => props.title === 'settings'
      );
      fireEvent.press(headerButton);
      expect(navigate).toHaveBeenCalledWith('Settings');
    });

    test('should handle headerRight navigate to ProfileHelp', async () => {
      const navigate = jest.fn();
      const navData = { navigation: { navigate } };

      await act(async () => {
        component = create(
          profileScreenOptions(navData).headerRight({
            tintColor: 'white',
          })
        );
      });

      const headerButton = component.root.find(
        ({ props }) => props.title === 'help'
      );
      fireEvent.press(headerButton);
      expect(navigate).toHaveBeenCalledWith('ProfileHelp');
    });

    test('tabBarIcon', () => {
      const tabBarIcon = profileScreenOptions().tabBarIcon({
        color: 'white',
      });
      expect(tabBarIcon).toMatchSnapshot();
    });
  });
});
