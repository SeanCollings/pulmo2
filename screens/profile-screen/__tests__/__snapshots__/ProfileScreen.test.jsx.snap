// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileScreen - unit test profileScreenOptions headerRight 1`] = `
<View
  style={
    Object {
      "flexDirection": "row",
    }
  }
>
  <HeaderButtons
    HeaderButtonComponent={[Function]}
    left={false}
  >
    <Item
      color="white"
      iconName="settings"
      onPress={[Function]}
      title="settings"
    />
  </HeaderButtons>
  <HeaderButtons
    HeaderButtonComponent={[Function]}
    left={false}
  >
    <Item
      color="white"
      iconName="help-circle-outline"
      onPress={[Function]}
      title="help"
    />
  </HeaderButtons>
</View>
`;

exports[`ProfileScreen - unit test profileScreenOptions headerTitle 1`] = `
<Text
  style={
    Object {
      "color": "white",
      "fontSize": 20,
    }
  }
>
  Profile
</Text>
`;

exports[`ProfileScreen - unit test profileScreenOptions options 1`] = `
Object {
  "headerRight": [Function],
  "headerTitle": [Function],
  "tabBarIcon": [Function],
}
`;

exports[`ProfileScreen - unit test profileScreenOptions tabBarIcon 1`] = `
<Icon
  allowFontScaling={false}
  color="white"
  name="account"
  size={26}
/>
`;

exports[`ProfileScreen - unit test should render 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "white",
      "flex": 1,
      "paddingTop": 10,
    }
  }
>
  <View
    style={
      Object {
        "flexDirection": "row",
        "paddingVertical": 5,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#3e4b57",
          "fontFamily": "tit-light",
          "fontSize": 17,
          "opacity": 0.8,
        }
      }
    >
      Work Average deviation: 17.64%
    </Text>
    <View
      style={
        Object {
          "justifyContent": "center",
          "paddingLeft": 5,
        }
      }
    >
      <Text
        allowFontScaling={false}
        style={
          Array [
            Object {
              "color": "#CCC",
              "fontSize": 24,
            },
            undefined,
            Object {
              "fontFamily": "material-community",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "paddingBottom": 10,
      }
    }
  >
    <View
      style={
        Object {
          "width": "100%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#002f56",
            "fontFamily": "tit-regular",
            "fontSize": 20,
            "textAlign": "center",
          }
        }
      >
        Selected level
      </Text>
      <View>
        <View
          style={
            Object {
              "alignSelf": "center",
              "flexDirection": "row",
              "justifyContent": "flex-start",
            }
          }
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "justifyContent": "center",
                "opacity": 1,
                "paddingTop": 4,
              }
            }
            testID="left"
          >
            <Text
              allowFontScaling={false}
              style={
                Array [
                  Object {
                    "color": "#CCC",
                    "fontSize": 30,
                  },
                  undefined,
                  Object {
                    "fontFamily": "material-community",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              ﬍
            </Text>
          </View>
          <View
            style={
              Object {
                "width": 80,
              }
            }
          >
            <RCTScrollView
              horizontal={true}
              keyboardShouldPersistTaps="always"
              pagingEnabled={true}
              scrollEnabled={false}
              showsHorizontalScrollIndicator={false}
            >
              <View>
                <View
                  style={
                    Object {
                      "justifyContent": "center",
                      "width": 80,
                    }
                  }
                >
                  <Text
                    style={
                      Object {
                        "color": "#3e4b57",
                        "fontFamily": "tit-regular",
                        "fontSize": 30,
                        "opacity": 1,
                        "textAlign": "center",
                      }
                    }
                  >
                    0
                  </Text>
                </View>
                <View
                  style={
                    Object {
                      "justifyContent": "center",
                      "width": 80,
                    }
                  }
                >
                  <Text
                    style={
                      Object {
                        "color": "#3e4b57",
                        "fontFamily": "tit-regular",
                        "fontSize": 30,
                        "opacity": 1,
                        "textAlign": "center",
                      }
                    }
                  >
                    0.5
                  </Text>
                </View>
                <View
                  style={
                    Object {
                      "justifyContent": "center",
                      "width": 80,
                    }
                  }
                >
                  <Text
                    style={
                      Object {
                        "color": "#3e4b57",
                        "fontFamily": "tit-regular",
                        "fontSize": 30,
                        "opacity": 1,
                        "textAlign": "center",
                      }
                    }
                  >
                    1
                  </Text>
                </View>
                <View
                  style={
                    Object {
                      "justifyContent": "center",
                      "width": 80,
                    }
                  }
                >
                  <Text
                    style={
                      Object {
                        "color": "#3e4b57",
                        "fontFamily": "tit-regular",
                        "fontSize": 30,
                        "opacity": 1,
                        "textAlign": "center",
                      }
                    }
                  >
                    1.5
                  </Text>
                </View>
                <View
                  style={
                    Object {
                      "justifyContent": "center",
                      "width": 80,
                    }
                  }
                >
                  <Text
                    style={
                      Object {
                        "color": "#3e4b57",
                        "fontFamily": "tit-regular",
                        "fontSize": 30,
                        "opacity": 1,
                        "textAlign": "center",
                      }
                    }
                  >
                    2
                  </Text>
                </View>
                <View
                  style={
                    Object {
                      "justifyContent": "center",
                      "width": 80,
                    }
                  }
                >
                  <Text
                    style={
                      Object {
                        "color": "#3e4b57",
                        "fontFamily": "tit-regular",
                        "fontSize": 30,
                        "opacity": 1,
                        "textAlign": "center",
                      }
                    }
                  >
                    2.5
                  </Text>
                </View>
                <View
                  style={
                    Object {
                      "justifyContent": "center",
                      "width": 80,
                    }
                  }
                >
                  <Text
                    style={
                      Object {
                        "color": "#3e4b57",
                        "fontFamily": "tit-regular",
                        "fontSize": 30,
                        "opacity": 1,
                        "textAlign": "center",
                      }
                    }
                  >
                    3
                  </Text>
                </View>
                <View
                  style={
                    Object {
                      "justifyContent": "center",
                      "width": 80,
                    }
                  }
                >
                  <Text
                    style={
                      Object {
                        "color": "#3e4b57",
                        "fontFamily": "tit-regular",
                        "fontSize": 30,
                        "opacity": 1,
                        "textAlign": "center",
                      }
                    }
                  >
                    3.5
                  </Text>
                </View>
                <View
                  style={
                    Object {
                      "justifyContent": "center",
                      "width": 80,
                    }
                  }
                >
                  <Text
                    style={
                      Object {
                        "color": "#3e4b57",
                        "fontFamily": "tit-regular",
                        "fontSize": 30,
                        "opacity": 1,
                        "textAlign": "center",
                      }
                    }
                  >
                    4
                  </Text>
                </View>
                <View
                  style={
                    Object {
                      "justifyContent": "center",
                      "width": 80,
                    }
                  }
                >
                  <Text
                    style={
                      Object {
                        "color": "#3e4b57",
                        "fontFamily": "tit-regular",
                        "fontSize": 30,
                        "opacity": 1,
                        "textAlign": "center",
                      }
                    }
                  >
                    4.5
                  </Text>
                </View>
                <View
                  style={
                    Object {
                      "justifyContent": "center",
                      "width": 80,
                    }
                  }
                >
                  <Text
                    style={
                      Object {
                        "color": "#3e4b57",
                        "fontFamily": "tit-regular",
                        "fontSize": 30,
                        "opacity": 1,
                        "textAlign": "center",
                      }
                    }
                  >
                    5
                  </Text>
                </View>
              </View>
            </RCTScrollView>
          </View>
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "justifyContent": "center",
                "opacity": 1,
                "paddingTop": 4,
              }
            }
            testID="right"
          >
            <Text
              allowFontScaling={false}
              style={
                Array [
                  Object {
                    "color": "#CCC",
                    "fontSize": 30,
                  },
                  undefined,
                  Object {
                    "fontFamily": "material-community",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              ﬏
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "paddingBottom": 10,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#3e4b57",
          "fontFamily": "tit-regular",
          "fontSize": 18,
          "opacity": 0.8,
        }
      }
    >
      0 day streak
    </Text>
  </View>
  <RCTScrollView
    style={
      Object {
        "width": "100%",
      }
    }
  >
    <View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "white",
            "elevation": 5,
            "height": 50,
            "justifyContent": "center",
            "marginVertical": 2,
            "opacity": 1,
            "paddingHorizontal": "10%",
            "shadowColor": "#000",
            "shadowOffset": Object {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 0.25,
            "shadowRadius": 2.84,
          }
        }
        testID="allActivities"
      >
        <View
          style={
            Object {
              "flexDirection": "row",
              "justifyContent": "space-between",
              "width": "100%",
            }
          }
        >
          <View
            style={
              Object {
                "flexDirection": "row",
                "justifyContent": "center",
              }
            }
          >
            <View
              style={
                Object {
                  "justifyContent": "center",
                  "paddingRight": 20,
                  "paddingTop": 1,
                }
              }
            >
              <Text
                allowFontScaling={false}
                style={
                  Array [
                    Object {
                      "color": "#CCC",
                      "fontSize": 21,
                    },
                    undefined,
                    Object {
                      "fontFamily": "material-community",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    Object {},
                  ]
                }
              >
                
              </Text>
            </View>
            <Text
              style={
                Object {
                  "color": "#3e4b57",
                  "fontFamily": "tit-regular",
                  "fontSize": 18,
                  "justifyContent": "center",
                  "opacity": 1,
                }
              }
            >
              All activities
            </Text>
          </View>
          <View
            style={
              Object {
                "flexDirection": "row",
                "justifyContent": "center",
                "paddingTop": 1,
              }
            }
          >
            <Text
              allowFontScaling={false}
              style={
                Array [
                  Object {
                    "color": "#CCC",
                    "fontSize": 23,
                  },
                  undefined,
                  Object {
                    "fontFamily": "material-community",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              
            </Text>
          </View>
        </View>
      </View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "white",
            "elevation": 5,
            "height": 50,
            "justifyContent": "center",
            "marginVertical": 2,
            "opacity": 1,
            "paddingHorizontal": "10%",
            "shadowColor": "#000",
            "shadowOffset": Object {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 0.25,
            "shadowRadius": 2.84,
          }
        }
        testID="favourites"
      >
        <View
          style={
            Object {
              "flexDirection": "row",
              "justifyContent": "space-between",
              "width": "100%",
            }
          }
        >
          <View
            style={
              Object {
                "flexDirection": "row",
                "justifyContent": "center",
              }
            }
          >
            <View
              style={
                Object {
                  "justifyContent": "center",
                  "paddingRight": 20,
                  "paddingTop": 1,
                }
              }
            >
              <Text
                allowFontScaling={false}
                style={
                  Array [
                    Object {
                      "color": "#CCC",
                      "fontSize": 21,
                    },
                    undefined,
                    Object {
                      "fontFamily": "material-community",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    Object {},
                  ]
                }
              >
                
              </Text>
            </View>
            <Text
              style={
                Object {
                  "color": "#3e4b57",
                  "fontFamily": "tit-regular",
                  "fontSize": 18,
                  "justifyContent": "center",
                  "opacity": 1,
                }
              }
            >
              Favourites
            </Text>
          </View>
          <View
            style={
              Object {
                "flexDirection": "row",
                "justifyContent": "center",
                "paddingTop": 1,
              }
            }
          >
            <Text
              allowFontScaling={false}
              style={
                Array [
                  Object {
                    "color": "#CCC",
                    "fontSize": 23,
                  },
                  undefined,
                  Object {
                    "fontFamily": "material-community",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              
            </Text>
          </View>
        </View>
      </View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "white",
            "elevation": 5,
            "height": 50,
            "justifyContent": "center",
            "marginVertical": 2,
            "opacity": 1,
            "paddingHorizontal": "10%",
            "shadowColor": "#000",
            "shadowOffset": Object {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 0.25,
            "shadowRadius": 2.84,
          }
        }
        testID="calendar"
      >
        <View
          style={
            Object {
              "flexDirection": "row",
              "justifyContent": "space-between",
              "width": "100%",
            }
          }
        >
          <View
            style={
              Object {
                "flexDirection": "row",
                "justifyContent": "center",
              }
            }
          >
            <View
              style={
                Object {
                  "justifyContent": "center",
                  "paddingRight": 20,
                  "paddingTop": 1,
                }
              }
            >
              <Text
                allowFontScaling={false}
                style={
                  Array [
                    Object {
                      "color": "#CCC",
                      "fontSize": 21,
                    },
                    undefined,
                    Object {
                      "fontFamily": "material-community",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    Object {},
                  ]
                }
              >
                ףּ
              </Text>
            </View>
            <Text
              style={
                Object {
                  "color": "#3e4b57",
                  "fontFamily": "tit-regular",
                  "fontSize": 18,
                  "justifyContent": "center",
                  "opacity": 1,
                }
              }
            >
              Calendar
            </Text>
          </View>
          <View
            style={
              Object {
                "flexDirection": "row",
                "justifyContent": "center",
                "paddingTop": 1,
              }
            }
          >
            <Text
              allowFontScaling={false}
              style={
                Array [
                  Object {
                    "color": "#CCC",
                    "fontSize": 23,
                  },
                  undefined,
                  Object {
                    "fontFamily": "material-community",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
